### DISCUSSION TITLE
Is there any way to do masking with DPO?

### DISCUSSION THREAD

**Discussion by mashdragon** [axolotl-ai-cloud/axolotl#2226] [Q&A]
So, I'm well aware that a DPO dataset essentially ranks some responses over other responses.

I have created a dataset in which I've corrected portions of a model's output. I've set it up so the "chosen" responses are my corrected outputs and the "rejected" responses are the original outputs. The catch is, these outputs are not complete responses, and the outputs are exactly the same up until the correction.

I am wondering if there is a way for me to mask the dataset during fine-tuning so that the model doesn't "learn" to stop generating early, since the responses in my dataset are not complete (I want the model to keep going but learn that at this stage you should be doing this, not that). Also, since the outputs are identical up until the correction, I'm wondering what kind of effect this will have on DPO, or if the algorithm just cancels out the effect of the identical tokens.

Thank you!

**Comment by NanoCode012** [comment]
Hey, could you elaborate what you mean the outputs are not complete responses? Are the "correct" ones not complete as well?

**Reply by mashdragon** [reply]
Yes, the correct responses are incomplete. The reason for this is that generating the full correct response is very costly and time consuming (it involves correcting the LLM so it goes in a different direction at many points - assume the context is very long). Here's an example presented in diff format. Both the chosen and rejected responses would include everything up to the different section, with the red part in the rejected and the green part in the chosen.

This first correction would try to have it not start with an exclamation:

```diff
  <|begin_of_text|><|start_header_id|>system<|end_header_id|>
  
  You are a helpful AI assistant for travel tips and recommendations<|eot_id|><|start_header_id|>user<|end_header_id|>
  
  What is France's capital?<|eot_id|><|start_header_id|>assistant<|end_header_id|>
  
- Bonjour! The capital of France is Paris
+ The capital of France is Paris
```

And then maybe I would have a second training example for this interaction later on to correct it to begin with history rather than present travel tips:

```diff
  <|begin_of_text|><|start_header_id|>system<|end_header_id|>
  
  You are a helpful AI assistant for travel tips and recommendations<|eot_id|><|start_header_id|>user<|end_header_id|>
  
  What is France's capital?<|eot_id|><|start_header_id|>assistant<|end_header_id|>
  
  The capital of France is Paris. 
  
- **Bonus Travel Tip:** While planning
+ Paris was first considered
```

Is something like this possible? I find myself correcting the LLM in the same way many times. My intuition is that showing it exactly what it did wrong and exactly what it should have said in a localized way could lead to better fine tuning performance and much less work for me. I already have to do this work anyway when I want a very high quality generation, to keep the LLM "on track".

**Reply by mashdragon** [reply]
@NanoCode012 Do you think it would be possible to do what I want using the pre-tokenized dataset format (https://axolotl-ai-cloud.github.io/axolotl/docs/dataset-formats/tokenized.html)?

I see that I can manually specify the attention mask and labels here. However, since I want to train on partial responses, I am confused by the directive in the docs. I think my partial responses shouldn't include the EOS token, which contradicts what is suggested here:

> You must add BOS and EOS, and make sure that you are training on EOS by not setting its label to -100.

I'm wondering if this approach (ignoring all tokens other than the ones I want to reject/accept by setting their labels to -100) would work. I assume I would set the attention mask to 1 for all the prompt tokens, but I don't know which other tokens would need an attention mask of 1...

I also noticed that in `rl.py` we have this during dataset loading:

```py
            else:
                # If no `type` is provided, assume the dataset is already in the expected format with
                # "prompt", "chosen" and "rejected" already preprocessed
                split_datasets[i] = data_set
```

So, I'm also confused by this one. I am guessing this means I would make it something like this (mostly taken from the example)?

```
{"prompt": {"input_ids":[271,299,99],"attention_mask":[1,1,1],"labels":[271,-100,99]}, "chosen": {"input_ids":[87,227,8383,12],"attention_mask":[1,1,1,1],"labels":[87,227,8383,12]}, "rejected": {"input_ids":[87,227,8383,11],"attention_mask":[1,1,1,1],"labels":[87,227,8383,11]}}
```

Sorry for these "new user" questions. I'm still learning, hopefully you can see some of the effort above!, so some indication of whether this is on track would be very helpful. I'm still confused about the difference between -100 label tokens and a attention mask of 0. Thank you so much!

**Reply by NanoCode012** [reply]
Hey! 

> Sorry for these "new user" questions. I'm still learning

No worries, we're happy to help guide you.

Earlier, I was asking around to see if others had some idea on this issue.

One idea proposed was to omit the EOS during the processing of your data. This could _maybe_ tell the model that this isn't the end of the generation and to not incorrectly teach it to stop. You'll have to experiment and get back to us on this, maybe a simple LORA run to ensure that your training won't cause abrupt stops.

> I think my partial responses shouldn't include the EOS token, which contradicts what is suggested here:

The reason we had that message was to let the user's know that we are not adding those tokens and to handle it themselves, in your case, you may choose to omit them. To clarify, this data loading is only supported in SFT as of the moment.

>  I'm still confused about the difference between -100 label tokens and a attention mask of 0.

Label of -100 would mean that it's ignored during loss calculation. Attention_mask of 0 would be for ignoring padding tokens which you don't need to worry about. Attention_mask should just always be `[1]*len(input_ids)`.

> So, I'm also confused by this one. I am guessing this means I would make it something like this (mostly taken from the example)?

Nope, the "processing" here refers to formatting it into the proper template (for example, adding chatml tokens to prompt https://github.com/axolotl-ai-cloud/axolotl/blob/341b118c9cde966ca00d1920efd218c1062cd968/src/axolotl/prompt_strategies/dpo/chatml.py#L121-L124 ). In your case, you want to add the BOS & conversation tokens (if any).

---

One thing, if you're training this, could you also include good/complete samples? This should help prevent the model from biasing to cut-off responses.

**Reply by mashdragon** [reply]
Thank you so much for such a detailed answer! I am very thankful for the tips about how the pre-tokenized dataset format is only supported for SFT currently and about the attention mask, and also the suggestion to try with complete examples. I didn't think of that.

So to recap, I will modify the relevant "processing" transformation function in `src/axolotl/prompt_strategies/dpo/*` to exclude the EOS token from the `chosen` and `rejected` keys and see if that works, putting my dataset back into the regular format rather than pre-tokenized but also taking care to ensure the tokenization would be the same by, for example, adding whitespace if the last token had some at the end. In another experiment I'll also add the EOS/EOT token manually for complete samples and see if that is needed for the model not to stop early.

I will run these experiments and report back how they go!

**Reply by NanoCode012** [reply]
>  I will modify the relevant "processing" transformation function in src/axolotl/prompt_strategies/dpo/* to exclude the EOS token from the chosen and rejected keys and see if that works, putting my dataset back into the regular format rather than pre-tokenized

Right

>  but also taking care to ensure the tokenization would be the same by, for example, adding whitespace if the last token had some at the end.

What do you mean by this? I don't think you need to manually tokenizing it anymore, only change the formatting.

> In another experiment I'll also add the EOS/EOT token manually for complete samples and see if that is needed for the model not to stop early.

Yep!

**Reply by mashdragon** [reply]
> What do you mean by this? I don't think you need to manually tokenizing it anymore, only change the formatting.

I just mean I need to be careful about the characters at the very end of a "partial prompt". In trying to generate a "guided" LLM output by iteration, what I do is I cut off the LLM's output, change the direction of the output by changing the last tokens to match the correct direction, and then resume generation from there.

What I've noticed is that for the model I'm currently using ([nvidia/Llama-3.1-Nemotron-70B-Instruct](https://huggingface.co/nvidia/Llama-3.1-Nemotron-70B-Instruct)), where I cut off the model can make a difference. If I do not preserve the tokens, for example by cutting off whitespace at the end if the last token was `'}\n\n'`, sometimes that completely changes the continuation. If I don't get that last token correct and just end it with `'}'`, then the LLM often predicts an EOT token afterward! So I just need to be careful I don't change the existing tokens in my training data.

(Also thank you for confirming the other things!! :slightly_smiling_face:)

**Reply by NanoCode012** [reply]
Oh, that's an interesting detail for that model. Let us know how the experiment goes!

**Reply by mashdragon** [reply]
I know this was closed maybe to clean up the discussions but I would like to report that I have never been able to do this primarily because I never developed the infrastructure for easily tracking improvements and have no where near enough data to scale it right now. So it might still be effective, I have no idea.

I tried to train a model that outputs "dog" instead of "cat" tokens but I don't think my training dataset was what was required to get a good result, basically that I didn't include anything else in my dataset.
